### PR TITLE
CLOUDSTACK-7539: [S3] Parallel deployment makes reference count of a …

### PR DESCRIPTION
**Pull request by DaanHoogland**
…cache in nfs secondary staging store negative(-1)

old review request by Hiroki Ohashi that is planned for 4.5.2

 @bhaisaab, @karuturi, @wilderrodrigues please review

### CODE REVIEW

**Comment by bhaisaab** [comment]
While LGTM and travis seems to be happy about it, my java kung - {locks and synchroization} -fu skills are not to be trusted

File: engine/storage/cache/src/org/apache/cloudstack/storage/cache/manager/StorageCacheManagerImpl.java [hunk 1]
```diff
@@ -216,15 +223,82 @@ public boolean stop() {
...
-        DataObject objOnCacheStore = store.create(data);
+                    obj = objectInStoreMgr.findObject(data, store);
+                    st = obj.getState();
+                }
+
+                if (st == ObjectInDataStoreStateMachine.State.Ready) {
+                    s_logger.debug("there is already one in the cache store");
+                    DataObject dataObj = objectInStoreMgr.get(data, store);
+                    dataObj.incRefCount();
+                    return dataObj;
```
Comment (karuturi) [inline engine/storage/cache/src/org/apache/cloudstack/storage/cache/manager/StorageCacheManagerImpl.java]: we need to `lock.notifyAll();` before return.

**Comment by wido** [comment]
Since the comment of @karuturi says there is something wrong with the code I don't think we can say LGTM here?

I actually encountered this issue today on a running 4.5.2 deployment build from the 4.5 branch.

**Comment by DaanHoogland** [comment]
@wido, sounds like you mean to have it in but are not satisfied because of @karuturi 's remark. I will look into adding the notify at some place as I think it makes sense.

**Comment by DaanHoogland** [comment]
I found documentation on the net [1] that this is fine but I changed the flow anyway 
[1]http://stackoverflow.com/questions/7971946/in-java-return-value-within-synchronized-block-seems-like-bad-style-does-it-re

**Comment by karuturi** [comment]
Its not always safe. http://stackoverflow.com/a/14837457/201514
In our example, the threads unnecessarily wait for 1000 ms if the previous threads exited from a path which doesnt do a notify
updated code looks good to me. :+1:

**Comment by DaanHoogland** [comment]
figured something like that but had't noticed the wait <:}

@bhaisaab @wido are you fine like with this?

**Comment by wido** [comment]
I just tested it on  4.5 management server and my deployment errors are gone. I'm fine with it.

LGTM

**Comment by karuturi** [comment]
@DaanHoogland  are you pushing this to 4.5 branch as well?

**Comment by wido** [comment]
@karuturi @DaanHoogland @bhaisaab I suggest we merge it into 4.5 as well so that it makes 4.5.2

**Comment by DaanHoogland** [comment]
will do

**Pull request merged**